src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,272 @@
+            char * key        = static_cast<char *>(chip::Platform::MemoryAlloc(keySize + 1));
+            uint8_t * data    = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(source.mTextEntryies[i].mDataSize));
+
+            VerifyOrExit(keySize < kMdnsTextMaxSize, error = CHIP_ERROR_INVALID_ARGUMENT);
+            VerifyOrExit(key != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            VerifyOrExit(data != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            strncpy(key, source.mTextEntryies[i].mKey, keySize + 1);
+            entry.mDataSize = source.mTextEntryies[i].mDataSize;
+            memcpy(data, source.mTextEntryies[i].mData, entry.mDataSize);
+            entry.mKey  = key;
+            entry.mData = data;
+        }
+    }
+
+exit:
+    if (error != CHIP_NO_ERROR)
+    {
+        FreeTextEntries(dest);
+    }
+
+    return error;
+}
+
+} // namespace
+
+namespace chip {
+namespace Mdns {
+
+ServicePool::Entry::Entry() : mHasValue(false), mIsTombstone(false), mNodeId(0), mFabricId(0)
+{
+    mService.mTextEntryies  = nullptr;
+    mService.mTextEntrySize = 0;
+}
+
+ServicePool::Entry::Entry(Entry && rhs)
+{
+    MoveFrom(std::move(rhs));
+}
+
+ServicePool::Entry & ServicePool::Entry::MoveFrom(Entry && rhs)
+{
+    mService     = rhs.mService;
+    mNodeId      = rhs.mNodeId;
+    mFabricId    = rhs.mFabricId;
+    mHasValue    = rhs.mHasValue;
+    mIsTombstone = rhs.mIsTombstone;
+
+    rhs.mHasValue               = false;
+    rhs.mIsTombstone            = false;
+    rhs.mService.mTextEntryies  = nullptr;
+    rhs.mService.mTextEntrySize = 0;
+
+    return *this;
+}
+
+CHIP_ERROR ServicePool::Entry::Emplace(const MdnsService & service, uint64_t nodeId, uint64_t fabricId)
+{
+    CHIP_ERROR error;
+
+    if (mHasValue)
+    {
+        FreeTextEntries(&mService);
+    }
+    mService               = service;
+    mNodeId                = nodeId;
+    mFabricId              = fabricId;
+    mService.mTextEntryies = nullptr;
+    error                  = CopyTextEntries(&mService, service);
+
+    if (error == CHIP_NO_ERROR)
+    {
+        mHasValue    = true;
+        mIsTombstone = false;
+    }
+    return error;
+}
+
+CHIP_ERROR ServicePool::Entry::Emplace(MdnsService && service, uint64_t nodeId, uint64_t fabricId)
+{
+    if (mHasValue)
+    {
+        FreeTextEntries(&mService);
+    }
+    mService               = service;
+    mNodeId                = nodeId;
+    mFabricId              = fabricId;
+    service.mTextEntryies  = nullptr;
+    service.mTextEntrySize = 0;
+    mHasValue              = true;
+    mIsTombstone           = false;
+
+    return CHIP_NO_ERROR;
+}
+
+void ServicePool::Entry::Clear()
+{
+    if (mHasValue)
+    {
+        FreeTextEntries(&mService);
+    }
+    mHasValue    = false;
+    mIsTombstone = false;
+}
+
+ServicePool::Entry::~Entry()
+{
+    Clear();
+}
+
+ServicePool::Entry & ServicePool::FindAvailableSlot(uint64_t nodeId)
```

Comment:
By the way, is there a reason this is not using `NodeId` as the type for node ids?  Please file a followup to switch to that.